### ISSUE TITLE
Trailing spaces not allowed in indented code blocks

### ISSUE THREAD

**Issue by wouter-admiraal-sonarsource** [DavidAnson/markdownlint#190]
Hi,

First of all, thanks for this awesome tool 👍 !

I ran into an issue when using indented code blocks. Consider this Markdown snippet:

```md
# example.md

Code snippet example:

    package org.example;
    
    public class Lorem {
    
      Lorem() {
      
      }
      
      public void tryMe() {
        /* ... */
      }
    
    }

End of code snippet.
```

This is using the original, "standard" Markdown (which doesn't have fenced code blocks). In this code snippet, I have several blank lines to improve readability. This means, the snippet renders like so:

```
package org.example;

public class Lorem {

  Lorem() {
  
  }
  
  public void tryMe() {
    /* ... */
  }

}
```

If the empty lines were to be trimmed, it would render something like this:

```
package org.example;
```

```
public class Lorem {
```

```
  Lorem() {
```

```
  }
```

```
  public void tryMe() {
    /* ... */
  }
```

```
}
```

So, the empty lines need to be indented like the rest.

However, rule MD009 flags this is as incorrect, and fails with:

```text
example.md: 5: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 4]
example.md: 7: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 4]
example.md: 9: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 6]
example.md: 11: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 6]
...
```

I suggest this rule be deactivated whenever we're inside a code block. Lines containing nothing but white space probably have a special meaning when in context of code. I would imagine this could be done for both indented and fenced code blocks, but I could be wrong.

_Note: I actually started forking and was preparing a PR, but then I saw that you prefer an issue opened first :-). So, this is the issue. But JIC, [here is a commit](https://github.com/wouter-admiraal-sonarsource/markdownlint/commit/336267de51a2f3311b7cbb2213f7b0d3fdd8510c) with a proposed fix, along with an update test. The test -- correctly -- fails if the fix is not implemented._

**Comment by DavidAnson** [comment]
Thank you for the very detailed issue!

I agree with the change you propose for indented code blocks. However, I don’t see a need to change the behavior for fenced blocks. (Convince me with an example?)

Could you please create a PR with your commit? I propose the following tweaks:

- Change behavior to only apply for indented code blocks
- Move the condition of the new `if` statement into the existing conditional two lines below (should reduce code churn)
- Change the text of the test case to remove “newline” because it’s about trailing spaces, not CR/LF
- Send the PR against `next` branch (I need to update CONTRIBUTING, sorry)

**Comment by wouter-admiraal-sonarsource** [comment]
Thanks for the reply. I'm not sure for the fenced blocks. I was more thinking: "it won't hurt".

Looking through the code, I don't see any way to see in what kind of code block the current line is. This means I'd introduce some new state when looping over the lines. Is that correct?

PR #191 created.

**Comment by DavidAnson** [comment]
It should be easy. I’ll add a note to the PR. Thank you!

**Comment by DavidAnson** [comment]
Part of `0.15.0`.
